### DISCUSSION TITLE
Getting a Line Number From an Annotation

### DISCUSSION THREAD

**Discussion by Yalnegheimish** [pymupdf/PyMuPDF#1197] [Looking for help] [answered]
Hey guys,

I'm currently trying to make a change log app to help me organize the changes I need to make after receiving a report. To make my job easier, I'd also like to include the line number of the sentences I need to change/fix. 
Here's the draft code for the Annot extraction but I still haven't found a way to get the line numbers.
```
 for i in range(doc.pageCount):
    page = doc[i]
    change = ''
    original = ''
    for a in page.annots():
      change = a.info["content"]
      original = page.get_textbox(a.rect)
      row_cells = changelog.add_row().cells
      row_cells[0].text = str(page.number)
      if a.type[1] == "Caret":
        row_cells[1].text = "Add"
      if a.type[1] == "StrikeOut":
        row_cells[1].text = "Delete/Replace"
      if a.type[1] == "Text":
        row_cells[1].text = "Add/Replace"
      row_cells[2].text = original
      row_cells[3].text = change
```

Thanks :)

**Comment by JorjMcKie** [comment]
Not sure I understand:
The text in sticky note annotations is one string - with no line breaks, except the author inserted them.
```python
>>> doc=fitz.open()
>>> page=doc.new_page()
>>> a=page.add_text_annot((100,100), "line1\nline2")
>>> a
'Text' annotation on page 0 of <new PDF, doc# 1>
>>> a.info
{'content': 'line1\nline2', 'name': 'Note', 'title': '', 'creationDate': '', 'modDate': '', 'subject': '', 'id': 'fitz-A0'}
>>> a.info["content"]
'line1\nline2'
>>> print(a.info["content"])
line1
line2
>>> 
```
If the author did so, ok. If not, then there are none and the string may be arbitrarily long.
The text is not visible on the PDF page - only if you hover the mouse over the note symbol. Consequently, you cannot extract it using `annot.get_texbox(rect)` - no matter what you use for `rect`.

**Reply by Yalnegheimish** [reply]
I'm trying to get the line number of where the annotated/highlighted text is in reference to the page. For example, I would like to get which line the striked out "Sid nisi" is and the script would return 2. If I asked for "Curabitur tortor" I'll get 6.
<img width="774" alt="Screen Shot 2021-08-10 at 1 41 57 PM" src="https://user-images.githubusercontent.com/79389821/128853641-e70a92f5-1a8c-43ad-b7f8-c44e192afbf7.png">

**Comment by JorjMcKie** [comment, marked as answer]
Ah, that's better.
First, you need to get the bboxes of all lines on the page.
Second, take the bbox (annot.rect) of your annotation and look in which line's bbox it is contained. Some snippet like this:
```python
lines = []
for b in page.get_text("dict")["blocks"]:
    if b["type"] != 0: continue
    for l in b["lines"]:
        lines.append(l)
# now sort the lines - they may be in arbitrary order!
lines.sort(key=lambda l: (l["bbox"][3], l["bbox"][0]))  # ascending vertical, then horizontal
# for some annot do this:
for i, line in enumerate(lines):
    if annot.rect in fitz.Rect(line["bbox"]):
        print(annot, "contained in line %i" % i)
        found = True
        break
if not found:
    print("annot not found")
```

**Reply by Yalnegheimish** [reply]
I'm not sure I understand what you've done exactly. Please correct me if I'm wrong.

So you start iterating over the blocks inside the page; if the block is an image you skip it, otherwise, it's added to the list. Then you sorted the the list but what you mean by vertical and horizontal?
The last loop iterates over I and all of the collected lines of the same page. If the annotation is in the list of lines then it prints the annotation and the lines, otherwise it prints not found.

**Comment by JorjMcKie** [comment]
almost correct: **_"otherwise, it's lines are added to the list._**
The "bbox" of all blocks, lines, spans ... is a tuple `(x0, y0, x1, y1)` of rectangle coordinates: x0 is the left border coordinate, y1 is the bottom coordinate. So I sort by y1 top to bottom, then by x0 = left to right. This is valid for Western alphabets only of course. Adjust when / where necessary.

**Comment by Yalnegheimish** [comment]
I see. Thank you for your help. :)
